### PR TITLE
MTI-710 Driver for Dora

### PR DESCRIPTION
**Pull request by ChichengZZZ**
# An IMU Driver developed for Dora

## Description

IMU (Inertial Measurement Unit) is a sensor device that is used to measure and provide information about an object's acceleration, angular velocity (rotation rate), and sometimes its orientation with respect to the Earth's coordinate system. The IMU combines multiple sensors to provide a comprehensive set of data for understanding the motion and orientation of the object it is attached to.

What task does it do?
The primary task of an IMU is to capture and track the object's motion and orientation in 3D space. It is commonly used in various applications, including robotics, drones, virtual reality systems, motion capture, and navigation, where accurate motion data is crucial for performing specific tasks or providing valuable insights.

Which hardware is supported?
IMUs are available in various configurations, and the specific hardware components can vary based on the application's requirements. A typical IMU consists of the following sensors:

Accelerometer: Measures linear acceleration, allowing the IMU to detect changes in velocity and direction.

Gyroscope: Measures angular velocity, providing information about rotational motion.

### Installation

Please watch the readme.md of this driver to install

### Launch Step

Please watch the readme.md of this driver to launch.

### Development Guidelines

There is no special steps for this software.

### Output

Its output will be a dictionary type including acceleration, angular velocity and direction, etc.

### References

List any references that might be useful:

- https://github.com/dora-rs

## Pull Request TODO: 

- [ ] Add a new folder for the new imu.
- [ ] Add this README template into the folder as well as the source code.

### CODE REVIEW

File: vendors/imu/mpu6050/readme.md
Comment (haixuanTao) [inline vendors/imu/mpu6050/readme.md:1]: readme should be capitalized: `README.md`

**Comment by ChichengZZZ** [comment]
@haixuanTao **I have overwritten the previous problematic code according to your requirements, please review it, thank you**

File: vendors/imu/MTI-710/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+dora logs first-dataflow custom_node_1
+```
+### After that, the content accepted by dora will appear   
+
+**if you need to print the context to test the output:(Use the command line below)**  
+- **try:RUST_LOG=debug dora start ./your path.yml --attach --hot-reload**  
+- **dora logs (number) (node_name)** 
+
+## API
+Both `mtdevice` and `mtdef` are official code interfaces of producer.
```
Comment (haixuanTao) [inline vendors/imu/MTI-710/README.md]: Do you mind sharing a link or even better a git to where the `mtdevice.py` and `mtdef.py` comes from?

File: vendors/imu/MTI-710/op_1/dora_node.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+                        "angular_velocity_covariance": self.imu_msg.angular_velocity_covariance,
+                        "linear_acceleration": {
+                            "x": self.imu_msg.linear_acceleration.x,
+                            "y": self.imu_msg.linear_acceleration.y,
+                            "z": self.imu_msg.linear_acceleration.z,
+                        },
+                        "linear_acceleration_covariance": self.imu_msg.linear_acceleration_covariance,
+                    }
+                    # 将 imu_dict 转换为字节类型
+                    print(imu_dict)
```
Comment (haixuanTao) [inline vendors/imu/MTI-710/op_1/dora_node.py]: Can we remove unnecessary print?

File: vendors/imu/MTI-710/op_1/dora_node.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+                    # size_of_list = sys.getsizeof(self.imu_msg)
+                    # print(self.imu_msg, type(self.imu_msg))
+                    # print(f"List size in bytes: {size_of_list}")
+                    # 假设在另一个地方得到了 serialized_data 数据
+                    # 你可以将其转换回原来的 Python 对象
+                    # reconstructed_imu_dict = pickle.loads(serialized_data)
+                    # 打印还原后的属性字典
+                    # print(reconstructed_imu_dict)
+
+                    # reset default values
```
Comment (haixuanTao) [inline vendors/imu/MTI-710/op_1/dora_node.py]: Can we remove commented code that is not related to documentation?

File: vendors/imu/MTI-710/op_1/dora_node.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    def fill_from_Timestamp(self, o):
+        """Fill messages with information from 'Timestamp' MTData2 block.
+        """
+        pass
+
+    def fill_from_Status(self, o):
+        """Fill messages with information from 'Status' MTData2 block.
+        """
+        pass
+
```
Comment (haixuanTao) [inline vendors/imu/MTI-710/op_1/dora_node.py:160]: Can we remove this if it is not used?

File: vendors/imu/MTI-710/op_1/dora_node.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+                rotation_speed = rotation_angle * self.delta_q_rate
+                f = rotation_speed / sa_2
+                x, y, z = f * dqx, f * dqy, f * dqz
+                self.imu_msg.angular_velocity.x = x
+                self.imu_msg.angular_velocity.y = y
+                self.imu_msg.angular_velocity.z = z
+                self.imu_msg.angular_velocity_covariance = self.angular_velocity_covariance
+                self.pub_imu = True
+        except KeyError:
+            pass
```
Comment (haixuanTao) [inline vendors/imu/MTI-710/op_1/dora_node.py:140]: Why do we pass on a key error? Let's avoid having silent KeyError message.
Can we use `if` statement to check if the message contain certain keys.

File: vendors/imu/MTI-710/op_1/dora_node.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+
+    def fill_from_Acceleration(self, o):
+        """Fill messages with information from 'Acceleration' MTData2 block."""
+        self.pub_imu = True
+
+        # FIXME not sure we should treat all in that same way
+        try:
+            x, y, z = o['Delta v.x'], o['Delta v.y'], o['Delta v.z']
+        except KeyError:
+            pass
```
Comment (haixuanTao) [inline vendors/imu/MTI-710/op_1/dora_node.py:86]: Why do we have KeyError? Can we not use if statement?

File: vendors/imu/MTI-710/op_1/dora_node.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+        try:
+            x, y, z = o['accX'], o['accY'], o['accZ']
+        except KeyError:
+            pass
+        self.imu_msg.linear_acceleration.x = x
+        self.imu_msg.linear_acceleration.y = y
+        self.imu_msg.linear_acceleration.z = z
+        self.imu_msg.linear_acceleration_covariance = self.linear_acceleration_covariance
+
+    def fill_from_Angular_Velocity(self, o):
```
Comment (haixuanTao) [inline vendors/imu/MTI-710/op_1/dora_node.py:100]: Can we replace `o` by a more appropriate variable name.

File: vendors/imu/mpu6050/readme.md
Comment (ChichengZZZ) [reply vendors/imu/mpu6050/readme.md:1]: ok

File: vendors/imu/MTI-710/README.md (hunk 1 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/README.md]: no problem, I have add a link. Besides, although this link points to the ros2 file, there is no ros2 component in these two files, just pure python code.

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 2 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py]: ok

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 3 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py]: ok

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 7 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:100]: o means operation, I think it is understandable

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 4 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:160]: No, I thought that the data sent by the Xsens sensor includes the two headers of Timestamp and Status, and there must be corresponding functions, but after I optimized the original code, this program no longer need to process the values of these two variables separately in a certain module. But these two pieces of code cannot be deleted.

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 5 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:140]: In my comment, you should be able to see that the first three data of the sensor initialization cannot be used as normal data, so this part of the pass is just for the preparation of those three times. After the data is stable, I tested it without passing this code.

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 6 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:86]: In my comment, you should be able to see that the first three data of the sensor initialization cannot be used as normal data, so this part of the pass is just for the preparation of those three times. After the data is stable, I tested it without passing this code.

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 4 above)
Comment (haixuanTao) [reply vendors/imu/MTI-710/op_1/dora_node.py:160]: Make sense. Can you write a comment telling that this should not be deleted?

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 7 above)
Comment (haixuanTao) [reply vendors/imu/MTI-710/op_1/dora_node.py:100]: Not in the whole PR have you expressively stated that o means operation. You should change this variable name. 

Single character variable is a very practice and should be avoided.

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 6 above)
Comment (haixuanTao) [reply vendors/imu/MTI-710/op_1/dora_node.py:86]: Which comment?

Comment (haixuanTao) [reply vendors/imu/MTI-710/op_1/dora_node.py:86]: You should change this to:
```python
       if "Delta v.x" in o.keys():
            x, y, z = o['Delta v.x'], o['Delta v.y'], o['Delta v.z']
       elif ...
```
It would make things clearer

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 4 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:160]: The data sent by the Xsens sensor includes the two headers of Timestamp and Status, and there must be corresponding functions,I let the header find the corresponding data processing function by using the function find_handler_name(self, name), but after I optimized the original code from producer, this program no longer need to process the values of these two variables separately in a certain module. So these two functions are required in the form and these two pieces of code cannot be deleted.

Comment (haixuanTao) [reply vendors/imu/MTI-710/op_1/dora_node.py:160]: Can you write a comment in the code?

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 6 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:86]: Your suggestion is very pertinent. But, your modification will make it inconvenient to modify the code. This code is only the first generation and will be iterated later. I think that instead of entangled in the format, it is better to leave more room for later modification and optimization.

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 7 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:100]: This is my mistake, but I think it can be expressed as a formal parameter introduction, and I will add a note about o in the comment.

File: vendors/imu/MTI-710/op_1/dora_node.py (hunk 4 above)
Comment (ChichengZZZ) [reply vendors/imu/MTI-710/op_1/dora_node.py:160]: Okay, teacher, I didn't understand what you meant just now.

**Review by bobdingAI — APPROVED** [review]
